effectful-th/src/Effectful/TH.hs

```diff
@@ -0,0 +1,238 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE TemplateHaskell #-}
+-- | Template Haskell utilities.
+module Effectful.TH
+  ( -- * Generate functions for dynamic effects
+    makeSendFunctions
+  , makeSendFunctionsWithOptions
+  , makeSendFunctionFor
+  , Options
+  , defaultOptions
+  , setMakeFunction
+  , setMakeFunctionFor
+  , setToFunctionName
+
+    -- * Re-exports
+  , Effect
+  , send
+  ) where
+
+import Control.Monad (forM)
+import Effectful (Eff, Effect, (:>))
+import Effectful.Dispatch.Dynamic (send)
+import Data.Char (toLower)
+import qualified Data.Map as Map
+import Language.Haskell.TH
+import Language.Haskell.TH.Datatype
+import Language.Haskell.TH.Datatype.TyVarBndr
+
+-- | Options for the generation of send functions.
+--
+-- The available options are documented in their respective setters:
+--
+--  * 'setMakeFunction': Select for which data constructors functions are
+--    generated.
+--  * 'setToFunctionName': How data constructor names are mapped to function
+--    names.
+data Options = Options
+  { optionsMakeFunction :: Name -> Bool
+  , optionsToFunctionName :: String -> String
+  }
+
+-- | Default options used by 'makeSendFunctions'. Those are:
+--
+--  * Generate functions for all data constructors of the effect type.
+--  * The function's name is the one of the data constructor with the first
+--    letter converted to lower case.
+defaultOptions :: Options
+defaultOptions = Options
+  { optionsMakeFunction = const True
+  , optionsToFunctionName = \case
+    x : xs -> toLower x : xs
+    _ -> error "Empty constructor name"
+  }
+
+-- | Control which data constructor to generate functions for.
+--
+-- The function passed as a first argument is a predicate used to decide for
+-- each constructor name whether a function is generated or not.
+setMakeFunction :: (Name -> Bool) -> Options -> Options
+setMakeFunction f options = options { optionsMakeFunction = f }
+
+-- | A version of 'setMakeFunction' expecting a list of names for which
+-- functions are generated.
+setMakeFunctionFor :: [Name] -> Options -> Options
+setMakeFunctionFor ns = setMakeFunction (`elem` ns)
+
+-- | Controls how to map constructor names to the names of the functions that
+-- are generated.
+setToFunctionName :: (String -> String) -> Options -> Options
+setToFunctionName f options = options { optionsToFunctionName = f }
+
+-- | Generate functions for sending an effect operation to its handler.
+--
+-- For example,
+--
+-- > -- These extensions are needed for the definition of the effect.
+-- > {-# LANGUAGE DataKinds #-}
+-- > {-# LANGUAGE GADTs #-}
+-- > {-# LANGUAGE TypeFamilies #-}
+-- > -- These extensions are needed for the generation of the functions.
```

Comment:
You can check in TH whether these are enabled and complain otherwise, see https://github.com/well-typed/optics/blob/82f4b8c4957e927147c7ca61b86d9629d1894ea5/optics-th/src/Optics/TH/Internal/Utils.hs#L112 and how it's used.

Reply:
Oh, nice! I'll have a look.